### PR TITLE
Add short names for crds

### PR DESCRIPTION
**Pull request by supereagle**
Fix comments: https://github.com/istio/istio/pull/9731#issuecomment-443342094

### CODE REVIEW

**Comment by GregHanson** [comment]
please see discussion in my previous PR where I attempted to address this:
https://github.com/istio/istio/pull/6114

**Comment by GregHanson** [comment]
cc @rshriram @ayj @frankbu

**Comment by ayj** [comment]
> Can you offer up some insight on how short name collision is avoided in a Kubernetes cluster for clusterscoped CRDs?

(referencing the k8s link from https://github.com/istio/istio/pull/9731#issuecomment-443839913)

> shortnames can be used within different groups, like kubectl get svc.mygroup.io

> when unqualified by a group, shortnames follow server specified group priority, which resolves core api groups first.
> 
> this matches behavior of actual names (non-shortnames). You could define a resource in your group named services and the results would be the same

**Comment by sdake** [comment]
Thanks @ayj for the link.  Also, it is unclear we can activate short names in a helm `tiller` installed environment.  See https://github.com/istio/istio/issues/9884 for more details.

(I also feel this PR should be held)

/hold

**Comment by supereagle** [comment]
@ayj  Thanks for your link. K8s has provided the way to avoid short name conflicts, but Helm upgrade is really a problem. I think it is ok to hold this PR until #9884 is fixed.

**Comment by supereagle** [comment]
@sdake Sorry, a little busy these days, I will test this week.

**Comment by supereagle** [comment]
@sdake The short names can take effect based on #10562.

My test steps are based on https://github.com/istio/istio/issues/9884#issuecomment-441918440:

```
1. Run [tiller-hack](https://github.com/istio/tools/tree/master/tiller-hack) tool
2. Delete the istio-init-crd jobs (They will be created again to ensure the CRDs are updated when `helm upgrade`)
3. Apply the code change in this PR
4. Run `helm upgrade` to update CRDs
```

The second step is missing in you comment. It is necessary, otherwise only the configmaps will be updated but the CRDs. I'd like to put these steps in our docs, what about in [Upgrading Istio](https://istio.io/docs/setup/kubernetes/upgrading-istio/)?

**Comment by supereagle** [comment]
@sdake  PTAL, thx

**Comment by esnible** [comment]
I like some of the short names, such as `vs`.  `crbaccfg` is way to difficult to remember!  Consider only creating short aliases only if they are memorable and only for common Istio types.

**Comment by supereagle** [comment]
> Consider only creating short aliases only if they are memorable and only for common Istio types.

I have considered these 2 points when adding short names. At the same time, there are already some discussions in #9731.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 1]
```diff
@@ -121,6 +131,8 @@ spec:
     kind: ClusterRbacConfig
     plural: clusterrbacconfigs
     singular: clusterrbacconfig
+    shortNames:
+    - crbaccfg
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: this one is too complicated.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 2]
```diff
@@ -185,6 +201,8 @@ spec:
     kind: HTTPAPISpecBinding
     plural: httpapispecbindings
     singular: httpapispecbinding
+    shortNames:
+    - hasb
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: we probably don't need a short name for this.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 3]
```diff
@@ -206,6 +224,8 @@ spec:
     kind: HTTPAPISpec
     plural: httpapispecs
     singular: httpapispec
+    shortNames:
+    - has
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: we probably don't need a short name for this.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 4]
```diff
@@ -227,6 +247,8 @@ spec:
     kind: QuotaSpecBinding
     plural: quotaspecbindings
     singular: quotaspecbinding
+    shortNames:
+    - qsb
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: we probably don't need a short name for this.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 5]
```diff
@@ -1004,6 +1028,8 @@ spec:
     kind: RbacConfig
     plural: rbacconfigs
     singular: rbacconfig
+    shortNames:
+    - rbaccfg
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: too complicated.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 6]
```diff
@@ -248,6 +270,8 @@ spec:
     kind: QuotaSpec
     plural: quotaspecs
     singular: quotaspec
+    shortNames:
+    - qs
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: don't know if this is useful enough to be shortnamed.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 7]
```diff
@@ -14,6 +14,8 @@ spec:
     listKind: VirtualServiceList
     plural: virtualservices
     singular: virtualservice
+    shortNames:
+    - vs
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: this is common and would be used.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 8]
```diff
@@ -36,6 +38,8 @@ spec:
     listKind: DestinationRuleList
     plural: destinationrules
     singular: destinationrule
+    shortNames:
+    - dr
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: this is common and would be used.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 9]
```diff
@@ -58,6 +62,8 @@ spec:
     listKind: ServiceEntryList
     plural: serviceentries
     singular: serviceentry
+    shortNames:
+    - se
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: this is common and would be used.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 10]
```diff
@@ -79,6 +85,8 @@ spec:
     kind: Gateway
     plural: gateways
     singular: gateway
+    shortNames:
+    - gw
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: this is common and would be used.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 11]
```diff
@@ -100,6 +108,8 @@ spec:
     kind: EnvoyFilter
     plural: envoyfilters
     singular: envoyfilter
+    shortNames:
+    - ef
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: not sure how used this is.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 12]
```diff
@@ -1027,6 +1053,8 @@ spec:
     kind: ServiceRole
     plural: serviceroles
     singular: servicerole
+    shortNames:
+    - sr
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: this is common and would be used.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml [hunk 13]
```diff
@@ -1050,6 +1078,8 @@ spec:
     kind: ServiceRoleBinding
     plural: servicerolebindings
     singular: servicerolebinding
+    shortNames:
+    - srb
```
Comment (sdake) [inline install/kubernetes/helm/istio-init/files/crd-10.yaml]: this is common and would be used.

**Comment by sdake** [comment]
my recommendation would be to start small with the rules that are most often used from kubectl `gw, vs, se, dr` and add additional PRs as operators request them.

**Comment by ayj** [comment]
I agree with @sdake on being conservative here.

> ... some of them (vs,gw,dr) would come is quite handy for development.

New versions of kubectl with bash completion will auto-complete CRD names, e.g.

```
$ kubectl get v                   # <hit TAB>
validatingwebhookconfigurations.admissionregistration.k8s.io  volumeattachments.storage.k8s.io
virtualservices.networking.istio.io 

$ kubectl get vi                  # <hit TAG>
$ kubectl get virtualservices.networking.istio.io 
```

File: install/kubernetes/helm/istio-init/files/crd-10.yaml (hunk 9 above)
Comment (supereagle) [reply install/kubernetes/helm/istio-init/files/crd-10.yaml]: I have checked that these short names are not used by Kubernetes. If they are used by other tools, I think users should clearly specify the groups.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml (hunk 5 above)
Comment (supereagle) [reply install/kubernetes/helm/istio-init/files/crd-10.yaml]: rcfg?

**Comment by supereagle** [comment]
> my recommendation would be to start small with the rules that are most often used from kubectl `gw, vs, se, dr` and add additional PRs as operators request them.

I agree. Not all CRDs need short names, and it is hard to find suitable short names for some of them, only need to cover most frequently used.

**Comment by supereagle** [comment]
> I'd also strongly recommend starting a thread on discuss.istio.io about this topic. Naming is hard.

Have created the discussion

**Comment by linsun** [comment]
i agree with @sdake 's comment on the common ones, typing destinationrule and virtualservice is getting a little too tedious for me.  can we split this PR and get the common ones merged first?

File: install/kubernetes/helm/istio-init/files/crd-10.yaml (hunk 5 above)
Comment (yangminzhu) [reply install/kubernetes/helm/istio-init/files/crd-10.yaml]: FYI: This is deprecated by `ClusterRbacConfig` in 1.1 and will be removed probably in 1.2 or 1.3.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml (hunk 1 above)
Comment (yangminzhu) [reply install/kubernetes/helm/istio-init/files/crd-10.yaml]: For now, can we just use `rbac` for this? 
/cc @liminw

**Comment by frankbu** [comment]
If we only do the networking ones initially (i.e., `gw`, `vs`, `dr`, `se`), to be complete maybe should also consider the one new networking resource in 1.1, `sidecar` -> `sc`.

File: install/kubernetes/helm/istio-init/files/crd-10.yaml (hunk 5 above)
Comment (supereagle) [reply install/kubernetes/helm/istio-init/files/crd-10.yaml]: Prefer to not add this

**Comment by supereagle** [comment]
> to be complete maybe should also consider the one new networking resource in 1.1, `sidecar` -> `sc`

`sc` has been used for storageclasses by Kubernetes.

Currently, there are 2 criterions for CRDs selection:

* Is Frequently used
* There is a memorable and non-conflicted short name

Maybe completion is not a required criterion.

**Comment by supereagle** [comment]
Based on the discussion above, I will create another PR only includes `gw`, `vs`, `dr` and `se` for common ones. Is it ok? @sdake  @linsun @esnible @frankbu

**Comment by linsun** [comment]
@supereagle SG, please create it so we can get it in 1.1 :)

**Comment by supereagle** [comment]
> @supereagle SG, please create it so we can get it in 1.1 :)

Have created #11969, PTAL

**Comment by supereagle** [comment]
Close as #11969 has been merged.

/close

**Pull request closed without merging**